lib/plugins/threaded-signing/plugin.c

```diff
@@ -254,30 +281,34 @@ threaded_openssl_get_signature(sv_threaded_plugin_t *self,
   SignedVideoReturnCode status = SV_OK;
 
   g_mutex_lock(&self->mutex);
-  if (self->plugin_state == THREADED_SIGNING_HAS_SIGNATURE && self->signature_info) {
-    if (self->signature_info->signature_size > max_signature_size) {
+  if (self->output_buffer_idx > 0) {
+    if (self->output_buffer[0].size > max_signature_size) {
       // If there is no room to copy the signature, report zero size.
       *written_signature_size = 0;
     } else {
-      memcpy(signature, self->signature_info->signature, self->signature_info->signature_size);
-      *written_signature_size = self->signature_info->signature_size;
+      // Get the oldest signature
+      memcpy(signature, self->output_buffer[0].signature, self->output_buffer[0].size);
+      *written_signature_size = self->output_buffer[0].size;
+      // Move buffer
+      output_data_t tmp = self->output_buffer[0];
+      int i = 0;
+      while (self->output_buffer[i].signature != NULL) {
+        self->output_buffer[i] = self->output_buffer[i + 1];
+        i++;
+      }
+      self->output_buffer[i - 1] = tmp;
+      if (self->output_buffer_idx > 0) {
+        self->output_buffer_idx -= 1;
+      }
     }
     // Change state and mark as copied.
-    self->plugin_state = THREADED_SIGNING_WAITS_FOR_HASH_TO_SIGN;
     has_copied_signature = true;
-  } else if (self->plugin_state == THREADED_SIGNING_ERROR) {
+  } else if (self->output_buffer[0].signing_error) {
```

Comment:
I think you should check ```signing_error``` first.
You can only end up here if ```output_buffer_idx = 0```, then it does not make sense to check the content in position 0 right?

Also, even if there is an error you need to move the buffer.